In [1]:
import numpy as np
import pandas as pd


In [2]:
data = pd.read_csv("/content/mushrooms.csv")
data.isnull().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [6]:
data.head()

class cap-shape cap-surface  ... spore-print-color population habitat
0     p         x           s  ...                 k          s       u
1     e         x           s  ...                 n          n       g
2     e         b           s  ...                 n          n       m
3     p         x           y  ...                 k          s       u
4     e         x           s  ...                 n          a       g

[5 rows x 23 columns]

In [7]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
data = data.apply(LabelEncoder().fit_transform)   
data.head()

class  cap-shape  cap-surface  ...  spore-print-color  population  habitat
0      1          5            2  ...                  2           3        5
1      0          5            2  ...                  3           2        1
2      0          0            2  ...                  3           2        3
3      1          5            3  ...                  2           3        5
4      0          5            2  ...                  3           0        1

[5 rows x 23 columns]

In [5]:
data.describe()

class cap-shape cap-surface  ... spore-print-color population habitat
count   8124      8124        8124  ...              8124       8124    8124
unique     2         6           4  ...                 9          6       7
top        e         x           y  ...                 w          v       d
freq    4208      3656        3244  ...              2388       4040    3148

[4 rows x 23 columns]

## splitting into feature set and class label

In [8]:
x = data.drop(['class'] , axis = 1)
y = data['class']
# x = data.iloc[:, [2,3]].values
# y = data.iloc[:, 4].values

## splitting the dataset into training model

In [9]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.25, random_state = 0)

## Feature scaling

In [10]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)
x_train

array([[-0.84654963,  0.9504593 ,  1.79845965, ..., -1.09502697,
         1.07975868, -0.29613015],
       [ 1.03000635,  0.9504593 , -0.18431722, ..., -0.25551778,
         1.07975868, -0.29613015],
       [-0.84654963,  0.13732415, -0.5808726 , ..., -0.67527237,
        -2.88490904, -0.29613015],
       ...,
       [ 1.03000635,  0.13732415, -0.5808726 , ..., -0.25551778,
        -0.50610841, -0.29613015],
       [-0.84654963, -1.48894614, -0.18431722, ..., -0.25551778,
         0.28682513, -0.87555243],
       [ 1.03000635,  0.9504593 , -0.5808726 , ..., -0.25551778,
         1.07975868, -0.87555243]])

## Fitting Gaussian Naive Bayes to the training set

In [11]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(x_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

## Prediction of the test set result

In [12]:
y_pred = classifier.predict(x_test)

## making confusion matrix

In [13]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)
cm

array([[966,  95],
       [ 79, 891]])

## Fitting Bernoulli naive bayes to the trainning set

In [14]:
from sklearn.naive_bayes import BernoulliNB
classifier = BernoulliNB()
classifier.fit(x_train, y_train)

BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

# prediction of the test set result

In [15]:
y_pred = classifier.predict(x_test)

## making the confusion matrix

In [16]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)

0.897095027080256

## decision tree

In [17]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state = 0 , max_depth = 5)
dt.fit(x_train , y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=5, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [18]:
dt.score(x_train , y_train)

0.9783357951747907

In [19]:
predictions = dt.predict(x_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test , predictions)

0.982274741506647

# Random Forest Model

In [22]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(max_depth = 5)
rf.fit(x_train , y_train)
rf.score(x_train , y_train)

0.9896602658788775

In [23]:
predictions = rf.predict(x_test)
rf.score(x_test , y_test)


0.9916297390448056

# conclusion


  Bernoulli naive bayes - 89%
  Decision Tree Accuracy - 98.2%
  Random Forest Accuracy - 99%
  Hence, Ramdom Forest is comparitively better